codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakeException.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.websocketx;
+
+import io.netty.handler.codec.http.DefaultHttpResponse;
+import io.netty.handler.codec.http.HttpResponse;
+
+/**
+ * Client exception during handshaking process.
+ */
+public class WebSocketClientHandshakeException extends WebSocketHandshakeException {
+
+    private static final long serialVersionUID = 1L;
+
+    private final HttpResponse response;
+
+    public WebSocketClientHandshakeException(String message) {
+        this(message, null);
+    }
+
+    public WebSocketClientHandshakeException(String message, HttpResponse httpResponse) {
+        super(message);
+        if (httpResponse != null) {
+            response = new DefaultHttpResponse(httpResponse.protocolVersion(),
+                                               httpResponse.status(), httpResponse.headers());
+        } else {
+            response = null;
+        }
+    }
+
+    /**
+     * Returns a {@link HttpResponse response} if exception occurs during response validation otherwise {@code null}.
+     */
+    public HttpResponse response() {
```

Comment:
I wonder if it would be better to expose the protocol, status and headers directly here. This would make it clear that this `HttpResponse` is never a `FullHttpResponse` and so not needs to be released. WDYT ?

Reply:
I thought about this, but we have situations when an exception occurs before receiving a response and we do not have this data and all fields will be null, which is inconvenient (need to check before use) or we can throw the different type of exceptions with response data and  without it.

Reply:
got it... maybe we should just add to the java docs that this will never be `ReferenceCounted` and so no special treatment is needed.

Reply:
Make sense, even that WebSocket Handshake request/response in theory should not contain payload, but this is not explicitly prohibited by the RFC 😲

Reply:
Added summary